docs/snapshotting/design.md

```diff
@@ -0,0 +1,211 @@
+**Return value**: Will report success if the microVM already was or, as an effect of this command, becomes _Paused_.
+**Body**: None.
+
+### `PUT /resume`: will attempt to resume the microVM;
+**Prerequisites**: can be called at any time
+**Effects**:
+- _on success_: microVM is guaranteed to be _Running_;
+- _on failure_: no side-effects.
+
+**Return value**: will report success if the microVM already was or, as an effect of this command, becomes _Running_.
+**Body**: None.
+
+### `PUT /snapshot/create`: will attempt to create a full snapshot of the microVM
+**Prerequisites**: can be called only when the microVM state is _Paused_
+**Effects**:
+- _on success_:
+  - Two new files will be generated (with no side-effects to microVM state):
+    - the vmm snapshot - file containing device model state and emulation state, and
+    - the guest-mem full snapshot - file containing a full copy of the guest memory.
+  - The block devices contents are flushed to their backing files. At this point, the user should externally back these up.
+  - The generated snapshot files are immediately available to be used (current process releases ownership).
+  - Will reset the dirtied page bitmap, will mark all pages clean (from a incremental snapshot point of view).
+- _on failure_: No side-effects.
+regardless of success/failure: will not influence state, will not stop or end the microVM (it can be used as before).
+
+**Return value**: will report success if the snapshot files have been successfully generated, failure otherwise.
+**Body**:
+```yaml
+SnapshotCreateConfig:
+  type: object
+  required:
+    - snapshot_path
+    - mem_file_path
+  properties:
+    version:
+      type: integer
+      minimum: TBD
+      maximum: TBD
+      description: Optional field for snapshot format version. If omitted
+                   Firecracker will default to creating a snapshot with the
+                   current version.
+    snapshot_path:
+      type: string
+      description: Path to file which will be populated with the vmm state.
+    mem_file_path:
+      type: string
+      description: Path to file which will be populated with the guest memory.
+                   The file will contain a copy of the guest memory or copies
+                   of just the dirtied pages, depending on which URL this
+                   object is used.
+```
+
+### `PUT /snapshot/create-diff`: will attempt to create a "diff" snapshot of the microVM
+
+**Prerequisites**: can be called only when the microVM state is *Paused.*
+**Effects:**
+- _on success_:
+  - Two new files will be generated (with no side-effects to microVM state):
+    - the vmm snapshot - file containing device model state and emulation state, and
+    - the guest-mem diff snapshot - file containing a **diff copy** of the guest memory - the diff is consisted of the memory pages which have been dirtied since the last snapshot creation or since the creation of the microVM, whichever of these events was the most recent.
```

Comment:
Do you want to support diff across the resumes?

pause
create Snapshot1
resume
pause
create Snapshot2
resume
diff between Snapshot1 and Snapshot2

Which means we must track the dirty page in the whole lifetime of this VM since creating the first snapshot. If this is not the case, I think it should be specified.